### ISSUE TITLE
Project: implement the interactive charts block for topic pages (v2)

### ISSUE THREAD

**Issue by mlbrgl** [owid/owid-grapher#1447]
_Follow up to #1380_ 

## Core problem

We want to surface all the charts relevant to a topic on the corresponding topic page.

<img width="746" alt="Screenshot 2022-04-12 at 14 05 22" src="https://user-images.githubusercontent.com/13406362/162958910-ff914df1-3c2d-4ee6-8c64-a91ac62b5a9b.png">

## Status (July 9, 2022): ⚠️ blocked by technical text
- #1529 acted as a fallback for this v2. Work on the lightbox has been paused given the absence of content in the left pane: 
  - the list of related charts has been removed on the account of being confusing
  - "this chart is listed in the following articles" section is for now too infrequent
  - technical text is not available yet
- current progress on v2, as well as possible next steps once technical text is available is in #1524 ([comment](https://github.com/owid/owid-grapher/pull/1524#issuecomment-1179495342)) 

## Scope
`essential / important - now`
- [x] [All charts related to a topic are listed somewhere in the topic page](https://www.notion.so/owid/All-charts-related-to-a-topic-are-listed-somewhere-in-the-topic-page-40198c9e420b48148d249524bf155ff4)
- [ ] [visually differentiate key charts from non-key charts](https://www.notion.so/owid/visually-differentiate-key-charts-from-non-key-charts-4345f5f308904ada8d46be49b78730d8)
- [x] [“keyness” is an attribute of a chart in a particular topic](https://www.notion.so/owid/keyness-is-an-attribute-of-a-chart-in-a-particular-topic-afde1c75a1ca47c299b11357befa0abf)
  - [x] #1520  
- [ ] [on click, open chart in lightbox](https://www.notion.so/owid/on-click-open-chart-in-lightbox-3213d4eac1ee4a2b805a06b043f09b2e)
  - [x] #1498 
  - [ ] #1510

`essential / important - later`
- [x] [connect non-key charts to topics in the grapher admin](https://www.notion.so/owid/connect-non-key-charts-to-topics-in-the-grapher-admin-40ea71e09da645c9850883f5ada1418c): #1520 

## Support material
- [Author-centric comparison of connecting orphan charts to topics](https://github.com/owid/owid-grapher/issues/1447#issuecomment-1155383002)

## Versions

- v1: #1380
- v2-alpha1: [demo](https://github.com/owid/owid-grapher/issues/1447#issuecomment-1141176669), [technical plan](https://docs.google.com/presentation/d/1Wzu-HcKbTfqb2qgcMil5yCK8-efMtXR9CHDU5KB2-y0/edit#slide=id.p), [code](https://github.com/owid/owid-grapher/tree/1447-poc-interactive-charts-block-v2)
- v2-alpha2: [demo](https://github.com/owid/owid-grapher/issues/1447#issuecomment-1170993438), [code](https://github.com/owid/owid-grapher/pull/1524)
- v3: #1529

**Comment by mlbrgl** [comment]
_(originally submitted in https://github.com/owid/owid-grapher/issues/1380#issuecomment-1105188499)_

--

Below is a comparison of the different search architectures we could rely on: Algolia or local JS filtering on the page

| Features | with Algolia | without Algolia | 
|---|---|---|
| search through title, subtitle, tags and country | 👍 | 😕 | 
| can benefit from current site search configuration | 👍 | ⛔️ | 
| country selection from natural language | 👍 | 👍 | 
| search analytics | 👍 | ⛔️ | 
| potential for more unified UX between site search and chart search | 🙂 | 😐 | 
| `in the future` faceted search ([demo](https://www.notion.so/owid/Knowledge-graph-entries-as-topics-unify-grapher-and-Wordpress-taxonomies-4992c4f2671248c8872e19073debb172#6f6fc0142102463c81f890e7e9e49e45)) | 🙂 | 😟 (current topic only)|
| `in the future` search can expand up and down the topic hierarchy without leaving the page (same [demo](https://www.notion.so/owid/Knowledge-graph-entries-as-topics-unify-grapher-and-Wordpress-taxonomies-4992c4f2671248c8872e19073debb172#6f6fc0142102463c81f890e7e9e49e45)) | 🙂 | 🥵 | 

The table below describes how this particular block could be baked and served on page load, prior to any search. To the traditional UX and SEO advantage of SSR, there is here another dimension to take into consideration: Algolia's search quota.

| Baking scenarios | with Algolia | without Algolia |
|---|---|---|
| baking without SSR from Algolia index | 😡 means the block would be empty until a search is performed or the block would be populated lazily on scroll, hitting the search quota up to the number of page views on any given topic | N/A |
| baking with SSR from Algolia index | 😡 would likely burn through (public) search quota (200K): 20 bakes / day * 300 topics * 30 days = 180K (probably ok in the short term until we deploy these blocks at scale, but not desirable in the long term) | N/A | 
| baking with SSR from internal grapher DB | 😕 means there might be a discrepancy between what is listed at page load, and what is then returned upon search by Algolia (since the index is only refreshed weekly). As topic pages and their associated charts mature, this will be less of an issue) | 🙂 |

### Verdict
From my perspective, an Algolia implementation keeps more doors open for the future than a custom filtering widget and is my recommendation at this point. 

The original time investment is similar given the reliance on the existing search infrastructure (site search), and will likely be much less as we add more complex features. 

This comes at the cost of possible discrepancies between the initial page load and further search queries, which will gradually disappear as topics mature. One additional perspective is that in the absence of search queries, no discrepancy can and will be surfaced.

@danyx23 @larsyencken I'd be interested to hear your take on this

**Comment by mlbrgl** [comment]
Implementation intentions:

**1. use all our charts block as the source of truth for chart <-> topic(s) relationships** (_confidence: 90%_)
This is reversing some of the work in #1070 and #1455, where topics are defined in the grapher admin, as an attribute of the chart. The upshot of this approach are:
- conceptually and technically, this helps with disentangling grapher and Wordpress. Topics remain an artefact of the publication that grapher doesn't need to know about. From a product perspective, grapher can still ship with tag management for internal organisation, but these are not used for publishing on OWID.
- the relationship chart <-> topic can be ordered and defined in the same place
- there is natural support for multiple topics per chart, all the while allowing different ordering on every topic (a chart might be in the top 5 on one topic, but further down the list on another). Achieving this at the chart editor level seems more complex, and would probably require an additional "topic overview" screen to do the ordering (which diminishes the gain of having it in the admin).
- publishing a chart on an OWID topic is conceptually more understandable: instead of using tag that is shared between the entry and the chart (see ["All our charts" block](https://www.notion.so/owid/All-our-charts-block-126cf21d679341ae94269218373c0451)), authors will insert that chart onto the page directly, so there is less "magic" to contend with.

**2. use grid / card block for CMS container** (_confidence: 55%_)
- the grid block already supports ordering of cards through DnD and right / left arrows
- cards are versatile: we could imagine creating more bespoke presentations using non-chart cards for separations (similar to section slides in a slide deck)

**3. list all charts in SSG, then when clicking on "search", start Algolia with facets filtering exposed** (_confidence 95%_)
- see [rationale above](https://github.com/owid/owid-grapher/issues/1447#issuecomment-1129963059)
- see demo in [Searching and exploring the graph](https://www.notion.so/owid/Knowledge-graph-entries-as-topics-unify-grapher-and-Wordpress-taxonomies-4992c4f2671248c8872e19073debb172#b515bd0f9dad4c479f13626a120e30e0)

**Comment by mlbrgl** [comment]
Front-end prototype

https://user-images.githubusercontent.com/13406362/171004543-b515a7ec-063e-4a62-b904-e76c9f1d91a6.mp4

**Comment by mlbrgl** [comment]
Technical plan ([google slides](https://docs.google.com/presentation/d/1Wzu-HcKbTfqb2qgcMil5yCK8-efMtXR9CHDU5KB2-y0/edit#slide=id.p), [PDF](https://github.com/owid/owid-grapher/files/8851287/20220606-key_charts_technical_plan.pdf))

**Comment by mlbrgl** [comment]
An author-centric comparison of connecting charts to topics, either through grapher tags (current) or through [chart mentions](https://docs.google.com/presentation/d/1Wzu-HcKbTfqb2qgcMil5yCK8-efMtXR9CHDU5KB2-y0/edit#slide=id.g11cf5932813_0_24) in topic pages.


|                                | Current: grapher tags                                                                                                                                                                                              | Proposed: manual embed in content                                                                                          |
|--------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------|
| Connect a chart to a topic     | - go to /admin/posts and look up the page the chart needs to be attached to (this may be skipped if the author is familiar with the topic)                                                                         | - go to the relevant topic page                                                                                            |
|                                | - choose an existing tag among the page tags or create a new one                                                                                                                                                   | - copy the embed code and paste it in the content                                                                          |
|                                | - go to /admin/charts, locate the chart and tag accordingly                                                                                                                                                        |                                                                                                                            |
|                                |                                                                                                                                                                                                                    |                                                                                                                            |
| Move a chart to another topic  | - remove tag                                                                                                                                                                                                       | - go to the old topic page                                                                                                 |
|                                | - re-tag following the above procedure                                                                                                                                                                             | - cut the embed code                                                                                                       |
|                                |                                                                                                                                                                                                                    | - go to the new topic page                                                                                                 |
|                                |                                                                                                                                                                                                                    | - paste the embed code                                                                                                     |
|                                |                                                                                                                                                                                                                    |                                                                                                                            |
| Rename a topic                 | To maintain 1:1 coherence, go to /admin/tags and rename the associated tag too                                                                                                                                     | (nothing more to do)                                                                                                       |
|                                |                                                                                                                                                                                                                    |                                                                                                                            |
| Delete / merge / split a topic | Follow "Move a chart to another topic"                                                                                                                                                                             | Migrate charts to new / existing tag(s) (can this be done through the bulk FASTT editor?)                                  |
|                                |                                                                                                                                                                                                                    |                                                                                                                            |
| Main advantage                 | Charts can be tagged in the admin                                                                                                                                                                                  | Predictable: a chart is in the list if it has been pasted in                                                               |
|                                |                                                                                                                                                                                                                    |                                                                                                                            |
| Main inconvenient              | Coherence needs to be manually maintained by authors: grapher tags kept in sync with topic names, and grapher tags attached to pages (or reattached in case of deletion). Complex to reason with, prone to mistakes. | Orphan (non-key, non-article) charts need a home, most likely somewhere on the topic page. Copy / paste feels more manual. |

**Comment by mlbrgl** [comment]
`v2-alpha2`
Demo (feature complete):
- all features in [current scope](https://github.com/owid/owid-grapher/issues/1447#issue-1239904021)
- backend tagging in the admin is handled in #1520 


https://user-images.githubusercontent.com/13406362/176641716-672e3bba-9ae3-4564-8ee9-42001b1d0df9.mp4

This also ships with a refactor a the grapher page (see [rationale in technical plan](https://docs.google.com/presentation/d/1Wzu-HcKbTfqb2qgcMil5yCK8-efMtXR9CHDU5KB2-y0/edit#slide=id.g11cf5932813_0_72))

<img width="1440" alt="20220630-grapher_page_refactor-1447" src="https://user-images.githubusercontent.com/13406362/176644820-2f5616cf-7a32-46f6-9a28-c77b767f520c.png">

**Comment by mlbrgl** [comment]
Closing.

Due to the absence of technical text in the foreseeable future, the lightbox set-up designed for https://github.com/owid/owid-grapher/issues/1447 is blocked and has been postponed (and with it the implementation PR https://github.com/owid/owid-grapher/pull/1524).

As of July 2022, the version of the all charts block in production is https://github.com/owid/owid-grapher/pull/1529.

cc @mathisonian @larsyencken
